Adds MDX support and a Tab component for the different examples

alexandercarls opened the pull request:
- Added MDX Support while needing to reimplement some `remark`-Plugins.
- Reworked the Table of Contents due to the new data structure.
- Side effect: I also fixed ToC reloading on every link click when the page is first visited and react/gatsby routing didn't take place yet.
- Side effect: I fixed that the document headers are not hidden behind the top document bar for mobile and small desktop screensize.
- Added a granular `Tabs` component and a tailored `ExampleTabs` component for the domain with the fixed three different approaches.
- Copy-Toastr is not a faithful reimplementation. I like the "new" one better :). 

TODO after deployment (prod/staging):
- [x] Verify if the plugin `@darth-knoppix/gatsby-plugin-feed` still works with the `html` field, given that we now process mdx and feed reader look as clean as before.

Couldn't fix, waiting for new version:
- [ ] The article http://localhost:8000/docs/hotchocolate/api-reference/language/ currently throws the error `Invalid DOM property 'xmlnsXLink'. Did you mean 'xmlnsXlink'` from the generated mermaid graphic. This is supposedly fixed in mdx 2.0 https://github.com/mdx-js/mdx/issues/1148

michaelstaib commented:
Awesome, can't wait for this one :)

rstaib commented on website/src/components/misc/article-sections.tsx:
```diff
@@ -14,86 +14,99 @@ export const ArticleSections: FunctionComponent<ArticleSectionsProperties> = ({
-    let currentActiveId: string | undefined;
-
-    const handler = () => {
+    const headingPositions =
```
This name perhaps be simplified to `headings`.

alexandercarls replied:
Changed in 6722f3cf533d023def9e090f50c330edfc9296dc

rstaib commented on website/src/components/misc/article-sections.tsx:
```diff
@@ -14,86 +14,99 @@ export const ArticleSections: FunctionComponent<ArticleSectionsProperties> = ({
+      ((data.tableOfContents.items as TableOfContentItem[]) ?? [])
+        .flatMap(item => ([item, ...item.items ?? []]))
+        .map(item => {
+          const headingPosition = document.getElementById(item.url.substring(1))!
```
This could be named `position`.

alexandercarls replied:
Changed in 6722f3cf533d023def9e090f50c330edfc9296dc

rstaib commented on website/src/components/misc/article-sections.tsx:
```diff
@@ -14,86 +14,99 @@ export const ArticleSections: FunctionComponent<ArticleSectionsProperties> = ({
+
+          return {
+            id: item.url,
+            position: headingPosition,
```
Then we could remove `: headingPosition`. Anyway this part looks better now, probably because of mdx.

alexandercarls replied:
Changed in 6722f3cf533d023def9e090f50c330edfc9296dc.

The ToC field for the MDX type returned a proper data structure :)

rstaib commented:
Cannot wait to see it in action 👍

alexandercarls commented:
The tabs currently look like this:

Desktop:

![image](https://user-images.githubusercontent.com/4022053/104074709-14187180-5211-11eb-8f99-3a809d3c2948.png)

Mobile (iPhone SE):

![image](https://user-images.githubusercontent.com/4022053/104074777-3f9b5c00-5211-11eb-999b-0a74959b2f79.png)

PascalSenn commented:
@alex

> The tabs currently look like this:
> 
> Desktop:
> 
> ![image](https://user-images.githubusercontent.com/4022053/104074709-14187180-5211-11eb-8f99-3a809d3c2948.png)
> 
> Mobile (iPhone SE):
> 
> ![image](https://user-images.githubusercontent.com/4022053/104074777-3f9b5c00-5211-11eb-999b-0a74959b2f79.png)

this is really cool!!

PascalSenn commented:
we could even play around with a dropdown in the mobile version. this may scale better with more than three tabs

alexandercarls commented:
> we could even play around with a dropdown in the mobile version. this may scale better with more than three tabs

That certainly an improvement. I decided not to do it yet, as we currently don't have more than three tabs and `ExampleTabs` is tailored for the three different approached.

I would improve it when the scenario comes up :).

rstaib commented:
Very cool to see progress on the tabs component, just let me know when I could start the review.

alexandercarls commented:
@rstaib ready for review :)

fredericbirke commented:
I checked rss/atom/feed files and they are valid and working:
![image](https://user-images.githubusercontent.com/9481836/105630778-55cd2d00-5e4b-11eb-8be5-200e4f44d795.png)
They are even looking better than the current prod ones. Only thing we could improv are the headings for rss, the old ones link to the website if clicked, but this is more a nice to have.

There are still some issues, which all will be fixed in a follow up, because they provide the business value and are only small issies:
Tabs still don't look great:
![image](https://user-images.githubusercontent.com/9481836/105630855-a17fd680-5e4b-11eb-9230-901b7960577e.png)


Copy button is not fixed properly when scrolling:
![image](https://user-images.githubusercontent.com/9481836/105630871-c2482c00-5e4b-11eb-863b-b027981b11cb.png)


Tabs with text before code look bad:
![image](https://user-images.githubusercontent.com/9481836/105630883-cffdb180-5e4b-11eb-9004-35b3f835ce41.png)

fredericbirke reviewed (dismissed):
There are still some design issues, all described in this follow up:
https://github.com/ChilliCream/hotchocolate/issues/2930

Everything else is fine!

alexandercarls commented:
Added a new commit 7b770ceaabc22fa5e681c6a57be9e1e404fd6609 which fixed the horizontal scrolling in the pre tag for small viewports (i.e. language indicator and copy button no longer scroll) and added some margin when a tab begins with a paragraph.

fredericbirke reviewed (approved):
LGTM

PascalSenn commented:
/AzurePipelines Run

rstaib commented on website/src/components/articles/article-sections.tsx:
```diff
@@ -14,86 +15,92 @@ export const ArticleSections: FunctionComponent<ArticleSectionsProperties> = ({
-            ),
-          }}
-        />
+        <div onClick={handleCloseClick}>
```
Is there a reason for the extra div? Why not using the `onClick` directly on `TableOfContent`?

alexandercarls replied:
`TableOfContent` is not a styled-component but a normal React one. As far as I know you would need to explicitly declare `onClick` as a prop and drill that to the inner components container. That is not worth it in my opinion.

rstaib commented on website/src/components/articles/article-sections.tsx:
```diff
@@ -111,47 +118,38 @@ const Title = styled.h6`
+  padding: 0 25px 10px;
+  list-style-type: none;
+
+  @media only screen and (min-width: 1320px) {
```
Why not using the helper from Fred?

rstaib replied:
Not sure if there is a helper for `1320px` but maybe you could check this out :-)

alexandercarls replied:
Does not exist.

The breakpoints and helper can be improved. In my opinion they should be degrading and either desktop- or mobile first.
The currently used breakpoints are a wild mix between  `min-width` and `max-width`.

rstaib commented on website/src/components/mdx/auto-linked-heading.tsx:
```diff
@@ -0,0 +1,59 @@
+import { Link } from "gatsby";
+import React, { FC } from "react";
```
Would prefer `FunctionComponent` instead of `FC`. Consistency :-)

alexandercarls replied:
Changed. In future work I would like to use `FC`. That is less noisy and does not occupy almost the full print-width when component properties are also specified.

rstaib commented on website/src/components/mdx/auto-linked-heading.tsx:
```diff
@@ -0,0 +1,59 @@
+`;
+
+
+const H = styled.h1`
```
Would prefer not to go for abbreviations.

alexandercarls replied:
Changed to `Heading`

rstaib commented on website/src/components/mdx/code-block.tsx:
```diff
@@ -0,0 +1,195 @@
+    text-transform: uppercase;
+`;
+
+const RE = /{([\d,-]+)}/
```
Also here!

alexandercarls replied:
Changed to `matchHighlights`

rstaib commented on website/src/components/mdx/copy.tsx:
```diff
@@ -0,0 +1,95 @@
+import React, { FC, useState } from "react"
+import styled from "styled-components"
+
+const copyToClipboard = (content: string) => {
```
Here I would prefer a function.

alexandercarls replied:
Changed

rstaib reviewed (changes requested):
Good work! Just a few comments.

rstaib commented:
@fredericbirke at the moment we should not have any text in tabs. If so, then we should probably rewrite the section.

rstaib commented:
@fredericbirke can you point me to the file where you found text inside a tab!? thanks

alexandercarls commented on website/package.json:
```diff
@@ -23,12 +23,16 @@
     "@darth-knoppix/gatsby-plugin-feed": "^2.0.2",
+    "@mdx-js/mdx": "^1.6.22",
+    "@mdx-js/react": "^1.6.22",
+    "@types/prismjs": "^1.16.2",
```
Moved

alexandercarls commented on website/src/blog/2019-01-24-schema-stitching.md:
````diff
@@ -171,7 +171,7 @@ foo(id:123) {
 Since, we did not want to cram a query like this into one string we allow this to be done with a flat path.
 
-```
+```sdl
````
Correct. I changed it back to plain 'ol `text`

alexandercarls commented on website/src/blog/2019-01-24-schema-stitching.md:
````diff
@@ -307,7 +307,7 @@ HTTP directives let you decorate a schema SDL and thus let you map REST services
 Moreover, we will introduce a cast feature to our delegation path. This will basically allow you to use fragments without having to write the code.
 
-```
+```sdl
````
Same changes.

alexandercarls commented:
> @fredericbirke can you point me to the file where you found text inside a tab!? thanks

http://localhost:8000/docs/hotchocolate/defining-a-schema/unions-and-interfaces/#interface-definition

rstaib commented:
@fredericbirke we should rewrite that. But for now we keep it look ugly.

rstaib reviewed (changes requested):
Look good! Just one more thing :-D Could you please run `yarn format` so that prettier takes care of code styling.
